### PR TITLE
Test private interface methods

### PR DESCRIPTION
**Pull request by jerryxu158**
Also adds relevant Gradle scripting to only run the test when executing on a JDK 11+ VM

### CODE REVIEW

File: com.ibm.wala.core/src/testJava11/privateInterfaceMethods/testArrayReturn/returnArray.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package testArrayReturn;
+
+public class returnArray {
```
Comment (msridhar) [inline com.ibm.wala.core/src/testJava11/privateInterfaceMethods/testArrayReturn/returnArray.java:3]: We are trying to test private interface methods, but this is a class, not an interface.  Maybe try a test case exactly like what was in [the blog post](https://jakewharton.com/androids-java-9-10-11-and-12-support/)?
```java
interface Java9PrivateInterface {
  static String hey() {
    return getHey();
  }

  private static String getHey() {
    return "hey";
  }
}
```
Notice that `Java9PrivateInterface` is an `interface`, not a `class`.

Also, the capitalization in the class and method names here is not the standard style for Java.  Class names should start with a capital letter, while method names should be camel case.  See the naming guide [here](https://google.github.io/styleguide/javaguide.html#s5-naming).

File: .idea/encodings.xml [hunk 2]
```diff
@@ -1,4 +1,8 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
-  <component name="Encoding" addBOMForNewFiles="with NO BOM" />
```
Comment (msridhar) [inline .idea/encodings.xml:3]: We shouldn't be changing any `.idea` files as part of this PR.  So these changes should be removed

File: .idea/libraries-with-intellij-classes.xml [hunk 3]
```diff
@@ -0,0 +1,65 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (msridhar) [inline .idea/libraries-with-intellij-classes.xml:1]: Don't add this file as part of the PR

File: .idea/misc.xml [hunk 4]
```diff
@@ -7,7 +7,27 @@
   <component name="JavaScriptSettings">
     <option name="languageLevel" value="ES6" />
   </component>
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_8" default="true" project-jdk-name="1.8" project-jdk-type="JavaSDK">
+  <component name="MavenProjectsManager">
```
Comment (msridhar) [inline .idea/misc.xml:10]: remove these changes

File: com.ibm.wala.core/src/testJava11/privateInterfaceMethods/PrivateInterface.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package pricateInterfaceMethods;
```
Comment (msridhar) [inline com.ibm.wala.core/src/testJava11/privateInterfaceMethods/PrivateInterface.java:1]:
```suggestion
package privateInterfaceMethods;
```

File: com.ibm.wala.core/src/testJava11/testJava11.iml [hunk 6]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (msridhar) [inline com.ibm.wala.core/src/testJava11/testJava11.iml:1]: remove this file from the PR.  we don't want to commit this IDE config file

File: testJava11/src.iml [hunk 7]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (msridhar) [inline testJava11/src.iml:1]: remove this file from the PR.  we don't want to commit this IDE config file

**Review by msridhar — CHANGES_REQUESTED** [review]
This is a good start!  We still need to add the Gradle configuration to compile the test cases when running a build on JDK 11, and the WALA test code to actually check that things are working.  Will chat with you separately on how to get that going

File: .idea/encodings.xml (hunk 2 above)
Comment (jerryxu158) [reply .idea/encodings.xml:3]: to be honest, I don't know what these changes are nor how I triggered them

File: com.ibm.wala.core/src/test11Subjects/java/privateInterfaceMethods/testDoubleReturn/ReturnDouble.java [hunk 8]
```diff
@@ -0,0 +1,5 @@
+package privateInterfaceMethods.testDoubleReturn;
+
+import privateInterfaceMethods.PrivateInterface;
+
+public class ReturnDouble implements PrivateInterface {}
```
Comment (msridhar) [inline com.ibm.wala.core/src/test11Subjects/java/privateInterfaceMethods/testDoubleReturn/ReturnDouble.java:5]: I think we can delete this file; just having the `ReturnArray` test is good enough

File: com.ibm.wala.core/src/test11Subjects/java/privateInterfaceMethods/testDoubleReturn/testDoubleReturn.java [hunk 9]
```diff
@@ -0,0 +1,9 @@
+package privateInterfaceMethods.testDoubleReturn;
+
+public class testDoubleReturn {
```
Comment (msridhar) [inline com.ibm.wala.core/src/test11Subjects/java/privateInterfaceMethods/testDoubleReturn/testDoubleReturn.java:3]: Delete this file (see above)

File: com.ibm.wala.core/src/test11Subjects/java/privateInterfaceMethods/testIntReturn/HasPrivateInt.java [hunk 10]
```diff
@@ -0,0 +1,5 @@
+package privateInterfaceMethods.testIntReturn;
+
+import privateInterfaceMethods.PrivateInterface;
+
+public class HasPrivateInt implements PrivateInterface {}
```
Comment (msridhar) [inline com.ibm.wala.core/src/test11Subjects/java/privateInterfaceMethods/testIntReturn/HasPrivateInt.java:5]: Delete this file (see above)

File: com.ibm.wala.core/src/test11Subjects/java/privateInterfaceMethods/testIntReturn/testIntReturn.java [hunk 11]
```diff
@@ -0,0 +1,12 @@
+package privateInterfaceMethods.testIntReturn;
+
+public class testIntReturn {
```
Comment (msridhar) [inline com.ibm.wala.core/src/test11Subjects/java/privateInterfaceMethods/testIntReturn/testIntReturn.java:3]: Delete this file (see above)

File: com.ibm.wala.core/src/test11Subjects/java/privateInterfaceMethods/testArrayReturn/testArrayReturn.java [hunk 12]
```diff
@@ -0,0 +1,10 @@
+package privateInterfaceMethods.testArrayReturn;
+
+public class testArrayReturn {
```
Comment (msridhar) [inline com.ibm.wala.core/src/test11Subjects/java/privateInterfaceMethods/testArrayReturn/testArrayReturn.java:3]:
```suggestion
public class TestArrayReturn {
```

Will have to change elsewhere including `PrivateInterfaceMethodsTest`

File: com.ibm.wala.core/src/test11Subjects/java/privateInterfaceMethods/testStringReturn/HasPrivateInterface.java [hunk 13]
```diff
@@ -0,0 +1,5 @@
+package privateInterfaceMethods.testStringReturn;
+
+import privateInterfaceMethods.PrivateInterface;
+
+public class HasPrivateInterface implements PrivateInterface {}
```
Comment (msridhar) [inline com.ibm.wala.core/src/test11Subjects/java/privateInterfaceMethods/testStringReturn/HasPrivateInterface.java:5]: Delete this file (see above)

File: com.ibm.wala.core/src/test11Subjects/java/privateInterfaceMethods/testStringReturn/testStringReturn.java [hunk 14]
```diff
@@ -0,0 +1,9 @@
+package privateInterfaceMethods.testStringReturn;
+
+public class testStringReturn {
```
Comment (msridhar) [inline com.ibm.wala.core/src/test11Subjects/java/privateInterfaceMethods/testStringReturn/testStringReturn.java:3]: Delete this file (see above)

**Review by msridhar — CHANGES_REQUESTED** [review]
Looking good!  Few minor comments to address

File: com.ibm.wala.core/src/test11Subjects/java/privateInterfaceMethods/PrivateInterface.java [hunk 15]
```diff
@@ -0,0 +1,11 @@
+package privateInterfaceMethods;
+
+public interface PrivateInterface {//called RetT and GetT b/c it used to type T return/input
+  default <Object> void RetT(Object input) {
```
Comment (msridhar) [inline com.ibm.wala.core/src/test11Subjects/java/privateInterfaceMethods/PrivateInterface.java:4]:
```suggestion
  default void RetT(Object input) {
```

Method is no longer generic

File: com.ibm.wala.core/src/test11Subjects/java/privateInterfaceMethods/PrivateInterface.java [hunk 16]
```diff
@@ -0,0 +1,11 @@
...
+    GetT(input);
+  }
+
+  private <Object> Object GetT(Object input) {
```
Comment (msridhar) [inline com.ibm.wala.core/src/test11Subjects/java/privateInterfaceMethods/PrivateInterface.java:8]:
```suggestion
  private Object GetT(Object input) {
```

Method is no longer generic

**Review by msridhar — CHANGES_REQUESTED** [review]
Be sure to run `./gradlew goJF` to ensure your code is formatted correctly

File: com.ibm.wala.core/src/test11Subjects/java/privateInterfaceMethods/PrivateInterface.java [hunk 17]
```diff
@@ -0,0 +1,11 @@
+package privateInterfaceMethods;
+
+public interface PrivateInterface { // called RetT and GetT b/c it used to type T return/input
```
Comment (msridhar) [inline com.ibm.wala.core/src/test11Subjects/java/privateInterfaceMethods/PrivateInterface.java:3]:
```suggestion
public interface PrivateInterface {
```

File: com.ibm.wala.core/src/test/java/com/ibm/wala/core/tests/privateInterfaceMethods/PrivateInterfaceMethodsTest.java [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+import com.ibm.wala.types.ClassLoaderReference;
+import com.ibm.wala.types.MethodReference;
+import com.ibm.wala.types.TypeReference;
+import com.ibm.wala.util.CancelException;
+import java.io.IOException;
+import java.util.Collection;
+import org.junit.Assert;
+import org.junit.Test;
+
+// if necessary
```
Comment (msridhar) [inline com.ibm.wala.core/src/test/java/com/ibm/wala/core/tests/privateInterfaceMethods/PrivateInterfaceMethodsTest.java:35]:
```suggestion
```

**Review by msridhar — APPROVED** [review]
LGTM!

**Comment by msridhar** [comment]
@liblit would be great to get a review from you, particularly of the Gradle script changes

File: com.ibm.wala.core/build.gradle [hunk 19]
```diff
@@ -13,11 +13,21 @@ eclipse {
 	}
 }
 
+final boolean java11OrHigher = JavaVersion.current() >= JavaVersion.VERSION_11
 sourceSets {
 	testSubjects
+	if (java11OrHigher) {
+		test11Subjects
+	}
```
Comment (liblit) [inline com.ibm.wala.core/build.gradle:21]:
```suggestion
	test11Subjects
	if (!java11OrHigher) {
		test11Subjects.java.srcDirs = []
	}
```

File: com.ibm.wala.core/build.gradle [hunk 20]
```diff
@@ -55,6 +65,11 @@ dependencies {
 	testRuntimeOnly(
 			files(compileTestSubjectsJava.map { it.outputs.files.first() })
 	)
+	if (java11OrHigher) {
+		testRuntimeOnly(
+				files(compileTest11SubjectsJava.map { it.outputs.files.first() })
+		)
+	}
```
Comment (liblit) [inline com.ibm.wala.core/build.gradle:72]:
```suggestion
	testRuntimeOnly(
			files(compileTest11SubjectsJava.map { it.outputs.files.first() })
	)
```

File: com.ibm.wala.core/build.gradle [hunk 21]
```diff
@@ -13,11 +13,21 @@ eclipse {
...
 }
 
 final Provider<JavaCompile> compileTestSubjectsJava = tasks.named('compileTestSubjectsJava', JavaCompile)
+Provider<JavaCompile> compileTest11SubjectsJava = null
+if (java11OrHigher) {
+	compileTest11SubjectsJava = tasks.named('compileTest11SubjectsJava', JavaCompile)
+	compileTest11SubjectsJava.get().sourceCompatibility = JavaVersion.VERSION_11.toString()
+	compileTest11SubjectsJava.get().targetCompatibility = JavaVersion.VERSION_11.toString()
+}
```
Comment (liblit) [inline com.ibm.wala.core/build.gradle:30]:
```suggestion
tasks.named('compileTest11SubjectsJava', JavaCompile) {
	sourceCompatibility = JavaVersion.VERSION_11.toString()
	targetCompatibility = JavaVersion.VERSION_11.toString()
}
```

File: com.ibm.wala.core/build.gradle [hunk 22]
```diff
@@ -303,6 +318,23 @@ tasks.register('collectTestData', Jar) {
...
+//
+
+if (java11OrHigher) {
+	tasks.register('collectTestData11', Jar) {
+		archiveFileName.set 'com.ibm.wala.core.testdata11_1.0.0.jar'
+		from compileTest11SubjectsJava
+		includeEmptyDirs false
+		destinationDirectory = layout.buildDirectory.dir name
+	}
+}
```
Comment (liblit) [inline com.ibm.wala.core/build.gradle:335]:
```suggestion
tasks.register('collectTestData11', Jar) {
	archiveFileName.set 'com.ibm.wala.core.testdata11_1.0.0.jar'
	from compileTest11SubjectsJava
	includeEmptyDirs false
	destinationDirectory = layout.buildDirectory.dir name
}
```

File: com.ibm.wala.core/build.gradle [hunk 23]
```diff
@@ -338,6 +370,10 @@ tasks.named('processTestResources') {
 		from tasks.named(it)
 	}
 
+	if (java11OrHigher) {
+		from tasks.named('collectTestData11')
+	}
```
Comment (liblit) [inline com.ibm.wala.core/build.gradle:375]:
```suggestion
	from tasks.named('collectTestData11')
```

**Review by liblit — CHANGES_REQUESTED** [review]
The `build.gradle` changes proposed in this pull request add many `if (java11OrHigher)` conditionals scattered throughout the build script. I think we can collapse all of these down to just a single conditional. The idea is to arrange for the `test11Subjects` source set to be empty if not `java11OrHigher`. If we do that, then everything else can be unconditional. For example, the `compileTest11SubjectsJava` task simply won't do anything if it has no sources to compile. I've left suggested changes in this review to implement this suggestion. However, I admit that I have not actually tested them!

I've also suggested changes to how the `sourceCompatibility` and `targetCompatibility` properties of the `compileTest11SubjectsJava` task are set. The original pull request used `get()` to fetch the actual task, but that's undesirable: it forces the task to be instantiated regardless of whether it's actually needed for any given build. My suggested replacement leverages [task configuration avoidance APIs](https://docs.gradle.org/current/userguide/task_configuration_avoidance.html) to avoid configuring this task unless and until it's actually needed.

**Comment by msridhar** [comment]
Thanks for the great feedback, @liblit!  I've addressed it and added some further fixes and some cleanup enabled by your suggested changes

**Review by liblit — APPROVED** [review]

**Pull request merged**